openzwavemqtt/util/mqtt_client.py

```diff
@@ -0,0 +1,262 @@
+                self.create_client()  # reset connect/reconnect futures
+
+    async def subscribe_managers(self) -> None:
+        """Connect and subscribe to manager topics."""
+        async with AsyncExitStack() as stack:
+            # Keep track of the asyncio tasks that we create, so that
+            # we can cancel them on exit.
+            tasks = self.tasks = set()
+            stack.push_async_callback(cancel_tasks, tasks)
+
+            # Connect to the MQTT broker.
+            await stack.enter_async_context(self.asyncio_client)
+            # Reset the reconnect interval after successful connection.
+            self.reconnect_interval = 1
+
+            # Messages that doesn't match a filter will get logged and handled here.
+            messages = await stack.enter_async_context(
+                self.asyncio_client.unfiltered_messages()
+            )
+
+            for manager in self.managers.values():
+                task = asyncio.create_task(
+                    handle_messages(messages, manager.receive_message)
+                )
+                tasks.add(task)
+
+                # Note that we subscribe *after* starting the message loggers.
+                # Otherwise, we may miss retained messages.
+                topic = f"{manager.options.topic_prefix}#"
+                await self.subscribe(topic)
+
+            # Wait for everything to complete (or fail due to, e.g., network errors).
+            # Make sure we await new tasks added while awaiting the first tasks.
+            while self.tasks:
+                current_tasks = list(self.tasks)
+                self.tasks.clear()
+                await asyncio.gather(*current_tasks)
+
+    async def unsubscribe_manager(self, manager: OZWManager) -> None:
+        """Unsubscribe the manager topic."""
+        manager = self.managers[id(manager)]
+        topic = f"{manager.options.topic_prefix}#"
+        await self.unsubscribe(topic)
+
+    async def remove_manager(self, manager: OZWManager) -> None:
+        """Remove a manager."""
+        manager_id = id(manager)
+        manager = self.managers[manager_id]
+        await self.unsubscribe_manager(manager)
+        self.managers.pop(manager_id)
+        if not self.managers:
+            # Stop the client if there are no managers left.
+            assert self.client_task is not None
+            self.client_task.cancel()
+            try:
+                await self.client_task
+            except asyncio.CancelledError:
+                pass
+
+
+async def handle_messages(messages: Any, callback: Callable[[str, str], None]) -> None:
+    """Handle messages with callback."""
+    async for message in messages:
+        # Note that we assume that the message payload is an
+        # UTF8-encoded string (hence the `bytes.decode` call).
+        payload = message.payload.decode()
+        LOGGER.debug("Received message topic: %s, payload: %s", message.topic, payload)
```

Comment:
We are going to log this now once for each messenger, we should log it once. Maybe we can only log and process it when it matches the manager topic instead of forwarding each message to each manager ?

Reply:
This is resolved by using a single manager.